### PR TITLE
implement ArrayList

### PR DESCRIPTION
**Pull request by DmytroChernenko**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,125 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private int counter;
+
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:10]: remove empty lines

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,125 @@
...
+    private T[] storage = (T[]) new Object[DEFAULT_CAPACITY];
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:11]:
```suggestion
    private T[] values = (T[]) new Object[DEFAULT_CAPACITY];
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,125 @@
...
+        counter++;
+        for (int i = index; i < counter; i++) {
+            if (i == index) {
+                storage[i] = value;
+            } else {
+                T tmp2 = storage[i];
+                storage[i] = tmp;
+                tmp = tmp2;
+            }
+        }
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:38]: Screenshot 2022-07-21 at 17 24 38

**Review by Nick97-git — CHANGES_REQUESTED** [review]
Before you create PR with your homework, you need to go through the description of the task and checklist under the task, and correct all the points described there. The mentor will not check the work until the checklist items are corrected

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,113 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:6]: remove empty line

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,113 @@
...
     }
 
     @Override
     public void add(T value, int index) {
+        verifyIndexInBound(index);
+        checkExtention();
+        if (index == counter) {
+            add(value);
+            return;
+        }
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:25]: remove empty line

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,113 @@
...
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        verifyIndexInBoundOrEmpty(index);
+        T returnValue = values[index];
+
+        System.arraycopy(values, index + 1, values, index, counter - index - 1);
+
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:56]: remove empty lines

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,48 +1,113 @@
...
-
+        verifyIndexInBoundOrEmpty(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        verifyIndexInBoundOrEmpty(index);
+        T returnValue = values[index];
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:53]:
```suggestion
        T removedValue = values[index];
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,48 +1,113 @@
...
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < counter; i++) {
+            if (element != null && element.equals(values[i])) {
+                return remove(i);
+            } else if (element == null && values[i] == null) {
+                return remove(i);
+            }
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:68]: use one if

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,48 +1,113 @@
...
+        }
+        throw new NoSuchElementException();
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:70]: add informative message

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,48 +1,113 @@
...
+        return counter;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return counter == 0;
+    }
+
+    private void checkExtention() {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:83]:
```suggestion
    private void resize() {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,48 +1,113 @@
...
+        }
+    }
+
+    private void verifyIsNotEmpty() {
+        if (isEmpty()) {
+            throw new ArrayListIndexOutOfBoundsException("Arraylist is empty");
+        }
+    }
+
+    private void extend() {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:107]: remove this method and move its logic to the method where you use it

File: src/main/java/core/basesyntax/ArrayList.java [hunk 12]
```diff
@@ -1,48 +1,113 @@
...
+        }
+    }
+
+    private void verifyIndexInBoundOrEmpty(int index) {
+        if (isEmpty() || (index < 0 || index >= counter)) {
+            throw new ArrayListIndexOutOfBoundsException("incorrect index");
+        }
+    }
+
+    private void verifyIsNotEmpty() {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:101]: remove this method, you don't use it

File: src/main/java/core/basesyntax/ArrayList.java [hunk 13]
```diff
@@ -1,48 +1,113 @@
...
+        }
+    }
+
+    private void verifyIndexInBound(int index) {
+        if (index < 0 || index > counter) {
+            throw new ArrayListIndexOutOfBoundsException("incorrect index");
+        }
+    }
+
+    private void verifyIndexInBoundOrEmpty(int index) {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:95]:
```suggestion
    private void checkIndex(int index) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 14]
```diff
@@ -1,48 +1,113 @@
...
+        if (isEmpty() || (index < 0 || index >= counter)) {
+            throw new ArrayListIndexOutOfBoundsException("incorrect index");
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:97]:
```suggestion
            throw new ArrayListIndexOutOfBoundsException("Incorrect index: " + index);
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 15]
```diff
@@ -1,48 +1,113 @@
...
+        return counter == 0;
+    }
+
+    private void checkExtention() {
+        if (counter == values.length) {
+            extend();
+        }
+    }
+
+    private void verifyIndexInBound(int index) {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:89]: remove this method and move its logic directly to the method where you use it

**Review by Nick97-git — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java [hunk 16]
```diff
@@ -1,48 +1,98 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int counter;
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:7]:
```suggestion
    private int size;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 17]
```diff
@@ -1,48 +1,98 @@
...
+            T[] copy = values.clone();
+            double capacityMultiplier = 1.5;
+            values = (T[]) new Object[(int) (values.length * capacityMultiplier)];
+            System.arraycopy(copy, 0, values, 0, copy.length);
+        }
+    }
+
+    private void verifyIndexInBound(int index) {
+        if (index < 0 || index > counter) {
+            throw new ArrayListIndexOutOfBoundsException("incorrect index: " + index);
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:89]:
```suggestion
            throw new ArrayListIndexOutOfBoundsException("incorrect index: " + index + " for size " + size);
```

**Review by MaliukDaria — APPROVED** [review]

**Pull request closed without merging**